### PR TITLE
Added additional APIs to better support JNI on Spark

### PR DESCRIPTION
**Pull request by eisber**
Performance improvement for prediction is roughly 3-4x on our datasets.
Training data loading was also improved by avoiding memory copy. Performance measurements are still in progress.

### CODE REVIEW

File: src/c_api.cpp [hunk 1]
```diff
@@ -22,6 +22,7 @@
 #include <stdexcept>
 #include <mutex>
 #include <functional>
+// #include <chrono>
```
Comment (imatiach-msft) [inline src/c_api.cpp]: remove (?)

File: src/c_api.cpp [hunk 2]
```diff
@@ -1175,6 +1285,35 @@ int LGBM_BoosterPredictForCSR(BoosterHandle handle,
...
 
+int LGBM_BoosterPredictForCSRSingleRow(BoosterHandle handle,
+                              const void* indptr,
+                              int indptr_type,
+                              const int32_t* indices,
+                              const void* data,
+                              int data_type,
+                              int64_t nindptr,
+                              int64_t nelem,
+                              int64_t,
```
Comment (imatiach-msft) [inline src/c_api.cpp]: what is the variable here?

Comment (imatiach-msft) [reply src/c_api.cpp]: (int64_t,)

Comment (imatiach-msft) [reply src/c_api.cpp]: looks like this is num col, shouldn't it be used here (or removed in signature otherwise)

File: swig/lightgbmlib.i [hunk 3]
```diff
@@ -14,18 +14,152 @@
...
+                                     int64_t* out_len,
+                                     double* out_result) {
+    // Alternatives
+    // - GetIntArrayElements: performs copy
+    // - GetDirectBufferAddress: fails on wrapped array
+    // Some words of warning for GetPrimitiveArrayCritical
+    // https://stackoverflow.com/questions/23258357/whats-the-trade-off-between-using-getprimitivearraycritical-and-getprimitivety
+    
+    jboolean isCopy;
+    int* indices0 = (int*)jenv->GetPrimitiveArrayCritical(indices, &isCopy);
```
Comment (imatiach-msft) [inline swig/lightgbmlib.i]: this is very cool 👍

File: swig/lightgbmlib.i [hunk 4]
```diff
@@ -14,18 +14,152 @@
...
+  };
+  
+  int LGBM_DatasetCreateFromCSRDirect(JNIEnv *jenv,
+                                     jobjectArray arrayOfSparseVector,
+                                     int num_rows,
+                                     int64_t num_col,
+                                     const char* parameters,
+                                     const DatasetHandle reference,
+                                     DatasetHandle* out) {
+   jclass sparseVectorClass = jenv->FindClass("org/apache/spark/ml/linalg/SparseVector");
```
Comment (imatiach-msft) [inline swig/lightgbmlib.i]: seems a bit strange to have spark datastructures leak into lightgbm.  Might be good to have something like Spark in the name of the method to make it clear that this isn't even just for Java but an optimization for spark in particular.

File: src/c_api.cpp (hunk 2 above)
Comment (eisber) [reply src/c_api.cpp]: good point. I copied this from the function above. you want to remove it there too?

File: swig/lightgbmlib.i (hunk 4 above)
Comment (eisber) [reply swig/lightgbmlib.i]: fair point

File: src/c_api.cpp [hunk 5]
```diff
@@ -197,6 +197,41 @@ class Booster {
...
+        is_predict_leaf = true;
+      } else if (predict_type == C_API_PREDICT_RAW_SCORE) {
+        is_raw_score = true;
+      } else if (predict_type == C_API_PREDICT_CONTRIB) {
+        predict_contrib = true;
+      } else {
+        is_raw_score = false;
+      }
+
+        // TODO: config could be optimized away... (maybe using lambda callback?)
```
Comment (imatiach-msft) [inline src/c_api.cpp]: nit: align comment with code

File: src/c_api.cpp (hunk 2 above)
Comment (imatiach-msft) [reply src/c_api.cpp]: no, if it's copied I guess we should keep it then

File: src/c_api.cpp [hunk 6]
```diff
@@ -1175,6 +1285,35 @@ int LGBM_BoosterPredictForCSR(BoosterHandle handle,
...
+                              int64_t nindptr,
+                              int64_t nelem,
+                              int64_t,
+                              int predict_type,
+                              int num_iteration,
+                              const char* parameter,
+                              int64_t* out_len,
+                              double* out_result) {
+  API_BEGIN();
+  auto param = Config::Str2Map(parameter);
```
Comment (imatiach-msft) [inline src/c_api.cpp]: some of this code looks very similar to below, would be nice to refactor to a single common function

Comment (eisber) [reply src/c_api.cpp]: I agree it looks very much the same, but then again it's mostly boilerplate code that's replicated in other functions too (e.g. LGBM_BoosterPredictForFile). Maybe packaging up the Config creation and translation to OMP would be better. Different PR?

Comment (imatiach-msft) [reply src/c_api.cpp]: ok, ignore my comment then, this is fine.

**Comment by guolinke** [comment]
@StrikerRUS I think we can use these apis to improve the prediction in python package as well

**Pull request merged**